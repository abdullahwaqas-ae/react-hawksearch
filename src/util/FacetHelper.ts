import { Value } from "models/Facets";
import { HierarchicalValue } from "models/Facets/HierarchicalValue";

export function prepareHierarchicalFacetValues(facetValues: Value[]): null {
    debugger;
    const selectedValuesWithoutLevel = facetValues.map(v => v.Value && v.Value.substring(v.Value.indexOf("|") + 1));

    const selectedValue = !selectedValuesWithoutLevel && Object.keys(selectedValuesWithoutLevel).length > 0 ? selectedValuesWithoutLevel[0] : "";
    const selectedValuePath = selectedValue.split("/");

    facetValues.forEach(value => {
        if (!value.Value) {
            return;
        }

        let valueWithoutLevel = value.Value;
        valueWithoutLevel = valueWithoutLevel.substring(valueWithoutLevel.indexOf("|") + 1);
        const valuesArray = valueWithoutLevel.split("/");

        // handle expand 

        // end of handling expand
        let hierarchicalValueList: HierarchicalValue[];
        let childValuePath = "";
        valuesArray.forEach((childValue, level) => {
            //hierarchicalValueList.push(childValue);
            childValuePath = (childValuePath === "" ? childValue : (childValuePath + "/" + childValue));
            let hierarchicalValues = hierarchicalValueList.filter(v => v.ValuePath === (level + "|" + childValuePath));
            if (hierarchicalValues && Object.keys(hierarchicalValues).length > 0) {
                hierarchicalValueList = hierarchicalValues[0].Children;
                return;
            }

            let hierarchicalValue: HierarchicalValue = {
                Selected: false,
                Value: value.Value,
                ValuePath: value.Value || "",
                Label: value.Label,
                Count: value.Count,
                Level: level,
                Children: []
            };
            hierarchicalValueList.push(hierarchicalValue);
        });
    });
    return null;
}