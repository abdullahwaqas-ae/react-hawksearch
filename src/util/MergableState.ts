import { useState, Dispatch } from 'react';

type SetStateAction<T> = Partial<T> | ((prevState: T) => Partial<T>);

export function useMergableState<T>(
	initialValue: T,
	typeConstructor: new (init: Partial<T>) => T
): [T, Dispatch<SetStateAction<T>>] {
	const [state, setState] = useState(new typeConstructor(initialValue));

	function setStateAndMerge(value: SetStateAction<T>) {
		if (typeof value === 'function') {
			// if we're being passed a function, we're setting state in the form of setState(prevState => ...).

			setState(prevState => {
				// so we derive the new state from the previous state
				const newState = value(prevState);

				// and then set the new merged state
				return new typeConstructor({ ...prevState, ...newState });
			});

			return;
		}

		// otherwise, the new state was simply passed in
		setState(prevState => {
			// merge state together and set it
			return new typeConstructor({ ...prevState, ...value });
		});
	}

	return [state, setStateAndMerge];
}
