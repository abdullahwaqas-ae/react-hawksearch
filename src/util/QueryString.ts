import { Request } from 'models/Search/Request';

/** Represents parts of the browser query string that are fixed and are always single strings. */
interface ParsedQueryStringFixed {
	keyword?: string;
	sort?: string;
	pg?: string;
	mpp?: string;
	searchWithin?: string;
}

/**
 * Represents the parts of the browser query string that are dynamic (the selected facets). Facets
 * can have multiple values, so the value of these is always an array of strings.
 */
interface ParsedQueryStringDynamic {
	[key: string]: string[];
}

type ParsedQueryString = ParsedQueryStringFixed & ParsedQueryStringDynamic;

/**
 * Parses the input query string and returns an object that can be used to build a search request.
 * The object returned will usually have the keys: `keyword`, `sort`, `pg`, `mpp`, and then more keys
 * for every selected facet.
 * @param search The input query string.
 */
function parseQueryStringToObject(search: string) {
	const params = new URLSearchParams(search);

	const parsed: ParsedQueryString = {};

	params.forEach((value, key) => {
		if (key === 'keyword' || key === 'sort' || key === 'pg' || key === 'mpp' || key === 'searchWithin') {
			// `keyword` is special and should never be turned into an array
			parsed[key] = value;
		} else {
			// everything else should be turned into an array

			if (!value) {
				// no useful value for this query param, so skip it
				return;
			}

			// multiple selections are split by commas, so split into an array
			const multipleValues = value.split(',');

			// and now handle any comma escaping - any single value that contained a comma is escaped to '::'
			for (let x = 0; x < multipleValues.length; ++x) {
				multipleValues[x] = multipleValues[x].replace('::', ',');
			}

			parsed[key] = multipleValues;
		}
	});

	return parsed;
}

/**
 * Parses the input query string into a `HawkClient` client search request object.
 * @param search The input query string.
 */
export function parseSearchQueryString(search: string): Partial<Request> {
	const queryObj = parseQueryStringToObject(search);

	// extract out components, including facet selections
	const { keyword, sort, pg, mpp, searchWithin, ...facetSelections } = queryObj;

	return {
		Keyword: keyword,

		SortBy: sort,
		PageNo: pg ? Number(pg) : undefined,
		MaxPerPage: mpp ? Number(mpp) : undefined,

		SearchWithin: searchWithin,

		FacetSelections: facetSelections,
	};
}

/**
 * Converts a search query object (such as one returned from `parseSearchQueryString`) and converts
 * it into a browser query string
 * @param queryObj The query object to convert to a query string.
 */
function convertObjectToQueryString(queryObj: ParsedQueryString) {
	const queryStringValues: string[] = [];

	for (const key in queryObj) {
		if (queryObj.hasOwnProperty(key)) {
			if (key === 'keyword' || key === 'sort' || key === 'pg' || key === 'mpp' || key === 'searchWithin') {
				const value = queryObj[key];

				if (value === undefined || value === null) {
					// if any of the special keys just aren't defined or are null, don't include them in
					// the query string
					continue;
				}

				if (typeof value !== 'string') {
					throw new Error(`${key} must be a string`);
				}

				// certain strings are special and are never arrays
				queryStringValues.push(key + '=' + value);
			} else {
				const values = queryObj[key];

				// handle comma escaping - if any of the values contains a comma, they need to be escaped first
				const escapedValues: string[] = [];

				for (const unescapedValue of values) {
					escapedValues.push(unescapedValue.replace(',', '::'));
				}

				queryStringValues.push(key + '=' + escapedValues.join(','));
			}
		}
	}

	return '?' + queryStringValues.join('&');
}

/**
 * Converts a partial search request object into a browser query string.
 * @param searchRequest The search request object to convert.
 */
export function getSearchQueryString(searchRequest: Partial<Request>) {
	const searchQuery = {
		keyword: searchRequest.Keyword,

		sort: searchRequest.SortBy,
		pg: searchRequest.PageNo ? String(searchRequest.PageNo) : undefined,
		mpp: searchRequest.MaxPerPage ? String(searchRequest.MaxPerPage) : undefined,

		searchWithin: searchRequest.SearchWithin,

		...searchRequest.FacetSelections,
	} as ParsedQueryString;

	return convertObjectToQueryString(searchQuery);
}
