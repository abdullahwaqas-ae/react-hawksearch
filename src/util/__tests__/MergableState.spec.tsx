import React, { useEffect } from 'react';
import TestRenderer from 'react-test-renderer';

import { useMergableState } from 'util/MergableState';

describe('MergableState', () => {
	it('sets initial state', () => {
		// arrange
		function MergableState() {
			const [state] = useMergableState({
				initial: 'state',
			});

			return <div>{JSON.stringify(state)}</div>;
		}
		const renderer = TestRenderer.create(<MergableState />);

		// act & assert
		expect(renderer.toJSON()).toMatchSnapshot();
	});

	it('merges new state', () => {
		// arrange
		function MergableState() {
			const [state, setState] = useMergableState({
				initial: 'state',
			} as any);

			// act
			useEffect(() => {
				setState({
					otherState: 'other value',
				});
			}, []);

			return <div>{JSON.stringify(state)}</div>;
		}

		const renderer = TestRenderer.create(<MergableState />);

		// trigger an update so that the `useEffect` is triggered
		renderer.update(<MergableState />);

		// assert
		expect(renderer.toJSON()).toMatchSnapshot();
	});

	it('merges new state with function pattern', () => {
		// arrange
		function MergableState() {
			const [state, setState] = useMergableState({
				initial: 'state',
			} as any);

			// act
			useEffect(() => {
				setState(prevState => {
					return {
						otherState: 'other value',
					};
				});
			}, []);

			return <div>{JSON.stringify(state)}</div>;
		}

		const renderer = TestRenderer.create(<MergableState />);

		// trigger an update so that the `useEffect` is triggered
		renderer.update(<MergableState />);

		// assert
		expect(renderer.toJSON()).toMatchSnapshot();
	});

	it('merges existing state', () => {
		// arrange
		function MergableState() {
			const [state, setState] = useMergableState({
				initial: 'state',
			} as any);

			// act
			useEffect(() => {
				setState({
					otherState: 'other value',
					initial: 'should be different',
				});
			}, []);

			return <div>{JSON.stringify(state)}</div>;
		}

		const renderer = TestRenderer.create(<MergableState />);

		// trigger an update so that the `useEffect` is triggered
		renderer.update(<MergableState />);

		// assert
		expect(renderer.toJSON()).toMatchSnapshot();
	});

	it('merges existing state with function pattern', () => {
		// arrange
		function MergableState() {
			const [state, setState] = useMergableState({
				initial: 'state',
				counter: 1,
			} as any);

			// act
			useEffect(() => {
				setState(prevState => {
					return {
						otherState: 'other value',
						initial: 'should be different',
						counter: prevState.counter + 2,
					};
				});
			}, []);

			return <div>{JSON.stringify(state)}</div>;
		}

		const renderer = TestRenderer.create(<MergableState />);

		// trigger an update so that the `useEffect` is triggered
		renderer.update(<MergableState />);

		// assert
		expect(renderer.toJSON()).toMatchSnapshot();
	});
});
