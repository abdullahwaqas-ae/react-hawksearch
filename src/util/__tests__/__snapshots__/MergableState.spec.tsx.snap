// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MergableState merges existing state 1`] = `
<div>
  {"initial":"should be different","otherState":"other value"}
</div>
`;

exports[`MergableState merges existing state multiple times 1`] = `
<div>
  {"initial":"should really be different","otherState":"other value"}
</div>
`;

exports[`MergableState merges existing state with function pattern 1`] = `
<div>
  {"initial":"should be different","otherState":"other value","counter":3}
</div>
`;

exports[`MergableState merges new state 1`] = `
<div>
  {"initial":"state","otherState":"other value"}
</div>
`;

exports[`MergableState merges new state with function pattern 1`] = `
<div>
  {"initial":"state","otherState":"other value"}
</div>
`;

exports[`MergableState retains methods on the state object 1`] = `
Array [
  <div>
    {"initial":"should be different"}
  </div>,
  <div>
    function doSomething() {
          return this.initial;
        }
  </div>,
]
`;

exports[`MergableState sets initial state 1`] = `
<div>
  {"initial":"state"}
</div>
`;
